### PR TITLE
fix: Show a helpful error message when a subgraph does not return JSON or bad status code

### LINKED ISSUE #1183: Show a helpful error message when a subgraph does not return JSON
if the subgraph does not return JSON (possible example: #1163), we do not show a good error message, instead it will look like "response was malformed: expected value at line 1 column 1".

The subgraph service should instead:
- check that the `Content-Type` is JSON
- check that the status code  is 200

### PR DESCRIPTION
**Pull request by bnjjj**
close #1183 

+ Rename the `TowerSubgraphService` to `SubgraphService`

### CODE REVIEW

File: apollo-router/src/services/tower_subgraph_service.rs [hunk 1]
```diff
@@ -111,6 +111,18 @@ impl tower::Service<graphql::SubgraphRequest> for TowerSubgraphService {
 
             // Keep our parts, we'll need them later
             let (parts, body) = response.into_parts();
+            if let Some(content_type) = parts.headers.get(header::CONTENT_TYPE) {
+                if let Ok(content_type_str) = content_type.to_str() {
+                    // Using .contains because sometimes we could have charset included (example: "application/json; charset=utf-8")
+                    if !content_type_str.contains("application/json") {
+                        return Err(BoxError::from(graphql::FetchError::SubrequestHttpError {
+                            service: service_name.clone(),
+                            reason: format!("subgraph doesn't return JSON content (current content-type: {content_type:?})"),
```
Comment (o0Ignition0o) [inline apollo-router/src/services/tower_subgraph_service.rs:120]:
```suggestion
                            reason: format!("subgraph didn't return JSON (expected content-type: application/json; found content-type: {content_type:?})"),
```

File: apollo-router/src/services/tower_subgraph_service.rs [hunk 2]
```diff
@@ -122,6 +134,16 @@ impl tower::Service<graphql::SubgraphRequest> for TowerSubgraphService {
                         reason: err.to_string(),
                     }
                 })?;
+            if parts.status != StatusCode::OK {
+                return Err(BoxError::from(graphql::FetchError::SubrequestHttpError {
+                    service: service_name.clone(),
+                    reason: format!(
+                        "subgraph returns status error '{}': {})",
```
Comment (o0Ignition0o) [inline apollo-router/src/services/tower_subgraph_service.rs:141]:
```suggestion
                        "subgraph HTTP status error '{}': {})",
```

File: apollo-router/src/services/tower_subgraph_service.rs [hunk 3]
```diff
@@ -142,3 +164,126 @@ impl tower::Service<graphql::SubgraphRequest> for TowerSubgraphService {
...
+                .status(StatusCode::BAD_REQUEST)
+                .body(r#"BAD REQUEST"#.into())
+                .unwrap())
+        }
+
+        let make_svc = make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(handle)) });
+        let server = Server::bind(&socket_addr).serve(make_svc);
+        if let Err(e) = server.await {
+            eprintln!("server error: {}", e);
+        }
```
Comment (o0Ignition0o) [inline apollo-router/src/services/tower_subgraph_service.rs:200]: is this used in tests? if so should we just unwrap ? not quite sure what we re trying to do here.

File: apollo-router/src/services/tower_subgraph_service.rs [hunk 4]
```diff
@@ -142,3 +164,126 @@ impl tower::Service<graphql::SubgraphRequest> for TowerSubgraphService {
...
+    }
+
+    // starts a local server emulating a subgraph returning bad response format
+    async fn emulate_subgraph_bad_response_format(socket_addr: SocketAddr) {
```
Comment (o0Ignition0o) [inline apollo-router/src/services/tower_subgraph_service.rs:204]: do we plan on using both functions in tests only ? or do you wanna provide them to plugin users? not sure

**Review by o0Ignition0o — COMMENTED** [review]
Looks good overall, A couple of questions and suggestions (that i m not sure are worth it tbh 😅)

File: apollo-router/src/services/tower_subgraph_service.rs [hunk 5]
```diff
@@ -66,7 +66,7 @@ impl tower::Service<graphql::SubgraphRequest> for TowerSubgraphService {
             let body = serde_json::to_string(&body).expect("JSON serialization should not fail");
 
             let mut request = http::request::Request::from_parts(parts, body.into());
-            let app_json: HeaderValue = "application/json".parse().unwrap();
+            let app_json: HeaderValue = HeaderValue::from_static("application/json");
```
Comment (Geal) [inline apollo-router/src/services/tower_subgraph_service.rs:69]: :+1:

File: apollo-router/src/services/tower_subgraph_service.rs (hunk 3 above)
Comment (bnjjj) [reply apollo-router/src/services/tower_subgraph_service.rs:200]: As it's spawned in a tokio task we can't just panic, we need to log the error

File: apollo-router/src/services/tower_subgraph_service.rs [hunk 6]
```diff
@@ -122,6 +134,16 @@ impl tower::Service<graphql::SubgraphRequest> for TowerSubgraphService {
                         reason: err.to_string(),
                     }
                 })?;
+            if parts.status != StatusCode::OK {
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply apollo-router/src/services/tower_subgraph_service.rs]: I was also wondering that and apparently regarding specs no if it's a correct query it SHOULD always be `200` (https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md) But I can be less strict and accept all `2xx` what's best ?

**Comment by glasser** [comment]
FYI I noticed that this PR added some special support for `application/graphql+json`. Did we have specific users looking for this or was this based on reading the graphql-over-http spec? They've actually more recently updated the recommendation to use `application/graphql-response+json` instead, so we should probably support that in addition to or instead of `application/graphql+json`.  (And if we do more stuff with incoming `accept` headers as part of https://github.com/apollographql/router/issues/1882 then we might want to support `accept: application/graphql-requests+json` in incoming requests too, as suggested by the spec.)

**Comment by bnjjj** [comment]
@glasser the reason we added this special support was to support the spec. But apparently it changed. So definitely we need to update it

**Comment by abernix** [comment]
The specification isn't really finalized yet, so I think we can be careful about how quickly we lean into anything it currently "specifies".

**Comment by glasser** [comment]
agreed (esp re urgency) although i'll note that this particular change (to move away from a vague "it's a GraphQL thingy, not gonna tell you what it is, same mime type works for requests and responses despite their structure being unrelated, there's a mime type for requests even though there's no use case for it and it makes it harder to write servers" to this more specific type) did come from Apollo/me, so there's something to be said for Apollo eventually implementing our own suggestions :) Took me a few months to get around to implementing it in Apollo Server though :)

**Pull request merged**